pkg/kubelet/kuberuntime/kuberuntime_container_windows.go

```diff
@@ -0,0 +1,73 @@
+// +build windows
+
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kuberuntime
+
+import (
+	"github.com/docker/docker/pkg/sysinfo"
+
+	"k8s.io/api/core/v1"
+	kubeletapis "k8s.io/kubernetes/pkg/kubelet/apis"
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
+)
+
+// applyPlatformSpecificContainerConfig applies platform specific configurations to runtimeapi.ContainerConfig.
+func (m *kubeGenericRuntimeManager) applyPlatformSpecificContainerConfig(config *runtimeapi.ContainerConfig, container *v1.Container, pod *v1.Pod, uid *int64, username string) error {
+	config.Windows = m.generateWindowsContainerConfig(container, pod, uid, username)
+	return nil
+}
+
+// generateWindowsContainerConfig generates windows container config for kubelet runtime v1.
+// Refer https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/cri-windows.md.
+func (m *kubeGenericRuntimeManager) generateWindowsContainerConfig(container *v1.Container, pod *v1.Pod, uid *int64, username string) *runtimeapi.WindowsContainerConfig {
+	wc := &runtimeapi.WindowsContainerConfig{
+		Resources: &runtimeapi.WindowsContainerResources{},
+	}
+
+	cpuRequest := container.Resources.Requests.Cpu()
+	cpuLimit := container.Resources.Limits.Cpu()
+	if !cpuLimit.IsZero() {
+		cpuCount := int64(cpuLimit.MilliValue()+999) / 1000
+		cpuMaximum := 10000 * cpuLimit.MilliValue() / int64(sysinfo.NumCPU()) / 1000
+		if kubeletapis.ShouldIsolatedByHyperV(pod.Annotations) && cpuCount != 0 {
+			cpuMaximum = cpuLimit.MilliValue() / cpuCount * 10000 / 1000
+		}
+		// ensure cpuMaximum is in range [1, 10000].
+		if cpuMaximum < 1 {
+			cpuMaximum = 1
+		} else if cpuMaximum > 10000 {
+			cpuMaximum = 10000
+		}
+
+		wc.Resources.CpuCount = cpuCount
+		wc.Resources.CpuMaximum = cpuMaximum
+	}
+
+	cpuShares := milliCPUToShares(cpuLimit.MilliValue())
```

Comment:
Windows cpushares range from 1-10000, we need a different calculation than the one used by Linux. @darrenstahlmsft Any pointer to calculate cpushares?

Reply:
Is it just a simple proportion?  100m = 1000 cpu shares (1/10 in both cases)?  Interestingly, according to https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/resource-controls#cpu-shares-without-hyper-v-isolation , the number passed in (if not using hyper-v isolation) will be translated to a value from 1 through 9 (default 5) for weight on the JOBOBJECT_CPU_RATE_CONTROL_INFORMATION structure

Reply:
> Is it just a simple proportion? 100m = 1000 cpu shares (1/10 in both cases)? 

cpu shares = milliCPU * 1024 / 1000, so 100m = 102 cpu shares, and 10 cpus would be 10240 cpu shares (which exceeds 10000).

Will update the PR and scale cpuShares within [1,10000].

Reply:
The shares at the platform is only relative to other containers (of the same isolation type), and is not based on the host's resource maximums. If all containers are set to 1000 shares, they will all have (near) equal access to CPU when the system is loaded. If no value is specified, the default for Hyper-V isolation is 10. The default on Process isolation is 5000, which is only changeable if no other processor options are set on that container. Ie. CpuCount = 4, CpuShares = 10000 on a process isolated container will result in the CpuCount being used, and the shares being ignored (and defaulting to 5000).

I'm still attempting to find the guarantees regarding how the shares are applied relative to each other. I don't think we make any guarantees that jobs weighted at, for example, 1 and 2, would result in the job weighted at 2 having twice as much CPU, just that it will have higher priority access to the host's CPUs.

Reply:
@darrenstahlmsft So how do you suggest to calc shares? factor by 10 for hyperv and 5000 by process?